src/toil/server/wes/tasks.py

```diff
@@ -0,0 +1,279 @@
+# Copyright (C) 2015-2021 Regents of the University of California
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import json
+import logging
+import os
+import shutil
+import subprocess
+from typing import Dict, Any, List, Union
+
+from toil.server.celery_app import celery
+from toil.server.utils import (get_iso_time,
+                               get_file_class,
+                               link_file)
+
+logger = logging.getLogger(__name__)
+
+
+class ToilWorkflowRunner:
+    """
+    A class to represent a workflow runner to run the requested workflow.
+
+    Responsible for parsing the user request into a shell command, executing
+    that command, and collecting the outputs of the resulting workflow run.
+    """
+
+    def __init__(self, work_dir: str, request: Dict[str, Any], engine_options: List[str]):
+        self.work_dir = work_dir
+        self.request = request
+        self.engine_options = engine_options
+
+        self.wf_type: str = request["workflow_type"].lower().strip()
+        self.version: str = request["workflow_type_version"]
+
+        self.exec_dir = os.path.join(self.work_dir, "execution")
+        self.out_dir = os.path.join(self.work_dir, "outputs")
+
+        self.default_job_store = "file:" + os.path.join(self.work_dir, "toil_job_store")
+        self.job_store = self.default_job_store
+
+    def write(self, filename: str, contents: str) -> None:
+        with open(os.path.join(self.work_dir, filename), "w") as f:
+            f.write(contents)
+
+    def get_state(self) -> str:
+        with open(os.path.join(self.work_dir, "state"), "r") as f:
+            return f.read()
+
+    def set_state(self, state: str) -> None:
+        self.write("state", state)
```

Comment:
Having the one plain file that different processes write to and read from without any sort of locking makes me suspicious.

I *think* that `open(..., 'w')` will protect from writes splicing together; it makes a new directory entry, so the last person to open the file will win.

But it's possible for a reader to see a partial write; in particular, between the open call and the write call it seems quite likely that a reader could get a state of "".

If you really want to move atomically between states, you'd want to write the data to a temporary file in the same directory, and then rename it over the old one, which makes the new contant appear at that filename atomically.